src/v/cloud_storage/tests/topic_namespace_override_recovery_test.cc

```diff
@@ -0,0 +1,191 @@
+        for (int i = 0; i < records.size(); ++i) {
+            ASSERT_EQ(records[i].key, consumed_records[i].key);
+            ASSERT_EQ(records[i].val, consumed_records[i].val);
+        }
+
+        // Sync archiver, upload candidates (if needed) and upload manifest.
+        archiver.sync_for_tests().get();
+        std::ignore = archiver.upload_next_candidates().get();
+        archiver.upload_topic_manifest().get();
+    }
+
+    // Restart the cluster, wiping all local data in the process.
+    restart(should_wipe::yes);
+    RPTEST_REQUIRE_EVENTUALLY(5s, [this] {
+        return app.storage.local().get_cluster_uuid().has_value();
+    });
+
+    // Now, create a new topic namespace that differs from the original,
+    // but uses the original src_tp_ns as an override in topic_properties.
+    // By setting recovery to true, it will attempt to use the src topic for
+    // recovery upon creation.
+    model::topic_namespace dest_tp_ns{
+      model::kafka_namespace, model::topic{"tapioca"}};
+    model::ntp dest_ntp(dest_tp_ns.ns, dest_tp_ns.tp, pid);
+
+    // Expect that no requests will be made to the s3_imposter using the
+    // destination topic.
+    http_test_utils::response fail_response{
+      .body = "Should not have recieved a request with tapioca in it",
+      .status = http_test_utils::response::status_type::bad_request};
+    req_pred_t fail_dest_topic_request =
+      [](const http_test_utils::request_info& info) {
+          return info.url.contains("tapioca");
+      };
+    fail_request_if(fail_dest_topic_request, fail_response);
+
+    cluster::topic_properties dest_props;
+    dest_props.shadow_indexing = model::shadow_indexing_mode::full;
+    dest_props.retention_local_target_bytes = tristate<size_t>(1);
+    dest_props.remote_topic_namespace_override = src_tp_ns;
+    dest_props.recovery = true;
+
+    add_topic(dest_tp_ns, 1, dest_props).get();
+    wait_for_leader(dest_ntp).get();
+
+    // Assert consuming from recovered destination topic works fine.
+    kafka_consume_transport consumer(make_kafka_client().get());
+    consumer.start().get();
+    auto consumed_records
+      = consumer.consume_from_partition(dest_tp_ns.tp, pid, model::offset(0))
+          .get();
+
+    ASSERT_EQ(records.size(), consumed_records.size());
+    for (int i = 0; i < records.size(); ++i) {
+        ASSERT_EQ(records[i].key, consumed_records[i].key);
+        ASSERT_EQ(records[i].val, consumed_records[i].val);
+    }
+
+    // Produce more to the destination topic, s3_imposter will fail if any
+    // "tapioca" related requests are made.
+    auto partition = app.partition_manager.local().get(dest_ntp);
+    auto archiver_ref = partition->archiver();
+    ASSERT_TRUE(archiver_ref.has_value());
+    auto& archiver = archiver_ref.value().get();
+
+    tests::remote_segment_generator gen(make_kafka_client().get(), *partition);
+
+    ASSERT_EQ(
+      num_records_to_produce,
+      gen.num_segments(2)
```

Comment:
Probably worth fixing the interface since it's easy to make this mistake: num_segment produces until the total number of segments is >=2, rather than adding 2 segments. This value should be >=10 (worth double checking in your local runs that that's the case)

Reply:
Yeah, this was intended on my end- but if `10` is a safer value than `2`, I'm happy to update it.